src/lib.rs

```diff
@@ -105,45 +106,168 @@ pub mod visit;
 #[cfg(feature = "fold")]
 pub mod fold;
 
+////////////////////////////////////////////////////////////////////////////////
+
 #[cfg(feature = "parsing")]
-mod parsing {
-    use std::str::FromStr;
-
-    use super::*;
-    use synom::{Synom, ParseError};
-    use proc_macro2::TokenStream;
-
-    macro_rules! traits {
-        ($($ty:ident,)*) => ($(
-            impl From<TokenStream> for $ty {
-                fn from(stream: TokenStream) -> $ty {
-                    $ty::parse_all_unwrap(stream)
-                }
-            }
+pub use synom::ParseError;
 
-            impl FromStr for $ty {
-                type Err = ParseError;
+#[cfg(feature = "parsing")]
+use synom::{Synom, SynomBuffer};
 
-                fn from_str(s: &str) -> Result<Self, Self::Err> {
-                    $ty::parse_str_all(s)
-                }
+/// Parse tokens of source code into the chosen syn data type.
+///
+/// This is preferred over parsing a string because tokens are able to preserve
+/// information about where in the user's code they were originally written (the
+/// "span" of the token), possibly allowing the compiler to produce better error
+/// messages.
+///
+/// # Examples
+///
+/// ```rust,ignore
+/// extern crate proc_macro;
+/// use proc_macro::TokenStream;
+///
+/// extern crate syn;
+///
+/// #[macro_use]
+/// extern crate quote;
+///
+/// use syn::DeriveInput;
+///
+/// #[proc_macro_derive(MyMacro)]
+/// pub fn my_macro(input: TokenStream) -> TokenStream {
+///     // Parse the tokens into a syntax tree
+///     let ast: DeriveInput = syn::parse(input).unwrap();
+///
+///     // Build the output, possibly using quasi-quotation
+///     let expanded = quote! {
+///         /* ... */
+///     };
+///
+///     // Parse back to a token stream and return it
+///     expanded.parse().unwrap()
+/// }
+/// ```
+#[cfg(feature = "parsing")]
+pub fn parse<T: Synom>(tokens: proc_macro::TokenStream) -> Result<T, ParseError> {
```

Comment:
Can we make this take a trait object which is implemented both on `proc_macro::TokenStream` and `proc_macro2::TokenStream` so that it acts less confusingly? it would have a method to turn it into a `proc_macro2::TokenStream`. Perhaps `Into<proc_macro2::TokenStream>`?

That way we just DTRT, which could be really nice. This is especially the case when dealing with parsing code from outside of a procedural macro, where `parse<T>` wouldn't be usable, as you cannot create a `proc_macro::TokenStream` from outside of a procedural macro right now without triggering a panic.